phastSim/phastSim.py

```diff
@@ -641,23 +809,196 @@ def sampleMutation(self, allele, rates, rand):
+            node.genomePos = [(self.n_insertions, pos_left), (self.n_insertions, pos_right)]
+            return node
+
+
+    def sampleInsertion(self, rate, node, rand, level):
+
+        # sample a geometric random variable
+        insertion_length = 1 + int(np.log(rand)/np.log(rate))
+        
+        # TODO generate a nucleotide sequence using the correct model
+        insertion = np.random.choice(self.allelesList, insertion_length, self.insertionFrequencies)
+
+        print("*************", insertion)
+        # turn the nucleotide sequence into a genomeTree
+        genomeSeq = "".join(insertion)
+
+        newNode = genomeNode(level=level)
+
+        # generate the genome tree of this insert from (node, genomeSeq)
+        subTree = self.buildTree(newNode, genomeSeq, 0, len(genomeSeq) - 1)
+
+        self.normalizeRates(subTree, self.indels, self.codon, len(genomeSeq), self.scale)
+
+        # build a bigger tree of the form t = (node, subTree). 
+        newRoot = genomeNode(level=level)
+        newRoot.belowNodes = [node, subTree]
+        newRoot.rate = node.rate + subTree.rate
+
+
+        mutData = mutation(mType=mType.INS, genomePos=node.refNode.genomePos[0], insert=genomeSeq, index=self.n_insertions)
+
+        return mutData, newRoot
 
     def findPos(self, rand, parentGenomeNode, level):
         # find position to mutate along the genome, and update temporary genome tree structure as you go
         if parentGenomeNode.isTerminal:
             # reached a terminal node, now sample the mutation event at the position and update all rates
             node = parentGenomeNode.refNode
             a = parentGenomeNode.allele
+            
+            # we'll choose an indel if rand is less than a threshold determined by the insertion and deletion rates for this site
+            if self.indels:
+
+                # deletions
+                if rand < node.deletionRate:
+
+                    attempted_deletion_length = self.sampleDeletion(node.deletionLength, rand/node.deletionRate)
+                    pos = node.genomePos[0]
+                    parentGenomeNode.rate = node.rate
+                    mutEvent = mutation(mType=mType.DEL, genomePos=pos, length=attempted_deletion_length)
+                    # in the tidy up function (self.deleteNodes), the length of this mutEvent may be changed if we are in an edge case
+                    # where we delete at a site very close to the end of the genome. The rates for all nodes are also updated there. 
+                    return mutEvent
+                
+                # else insertion:
+                if rand < node.deletionRate + node.insertionRate:
+                
+                    mutEvent, mutEventRootNode = self.sampleInsertion(node.insertionLength, node, rand/(node.deletionRate + node.insertionRate), level)
+                    # TODO is this the right node to pass?! Check this!
+
+                    parentGenomeNode = mutEventRootNode
```

Comment:
I suspect there is a problem here: insertion events create and return a new node, the root of the created insertion sub-tree, but this node, from what I understand, is not "attached" to the current genome tree layer, that is, it's not reachable from its supposed parent.
In any case, even if I am wrong (I will have to go back and check), what is the purpose of this line here? parentGenomeNode is supposed to be the root of the current layer of the genome tree, so why is it replaced with the root of the sub-tree corresponding to the insertion event?

Reply:
Yes I think you are right - can we talk about how to fix this in our meeting?

Reply:
parentGenomeNode.belowNodes[i]=mutEventRootNode